### PR TITLE
[mlir][transform] Emit error message with `emitSilenceableFailure`

### PR DESCRIPTION
**Pull request by srcarroll**
The previous implementation used a `notifyMatchFailure` to emit failure message inappropriately and then used the `emitDefaultSilenceableFailure`.  This patch changes this to use the more appropriate `emitSilenceableFailure` with error message.  Additionally a failure test has been added.

### CODE REVIEW

**Comment by srcarroll** [comment]
I'm still waiting on my local build to test this change

File: mlir/test/Dialect/Linalg/flatten-unsupported.mlir [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// RUN: mlir-opt %s -transform-interpreter -split-input-file -verify-diagnostics
+
+func.func @non_elementwise(%arg0: memref<2x3xf32>, %arg1: memref<3x4xf32>, %arg2: memref<2x4xf32>) {
+  // expected-error @+1 {{only elementwise flattening is supported}}
```
Comment (ftynse) [inline mlir/test/Dialect/Linalg/flatten-unsupported.mlir:4]: Nit: prefer `@below`

**Review by ftynse — APPROVED** [review]

**Comment by srcarroll** [comment]
As a general rule of thumb, should i even bother waiting for the windows buildkite to finish? often it just hangs for several hours, other times it just fails for reasons irrelevant to the PR, and rarely does it actually finish and pass (for me at least)

**Comment by srcarroll** [comment]
need to fix this one too https://github.com/llvm/llvm-project/pull/86146/files#diff-61bfc4a79549b25af3e46ff8002c9a9aea61fb537b565435ba0957e2b53b9adaR3286. just need to figure out a test case to emit this

**Comment by srcarroll** [comment]
> need to fix this one too https://github.com/llvm/llvm-project/pull/86146/files#diff-61bfc4a79549b25af3e46ff8002c9a9aea61fb537b565435ba0957e2b53b9adaR3286. just need to figure out a test case to trigger it

@ftynse this was done after your approval so i'll wait for you thumbs up on this

**Comment by ftynse** [comment]
> @ftynse As a general rule of thumb, should i even bother waiting for the windows buildkite to finish? often it just hangs for several hours, other times it just fails for reasons irrelevant to the PR, and rarely does it actually finish and pass (for me at least)

Use your best judgement and be ready to revert if submitting before it passes. I usually wait when changing the build system (Windows/MSVC tends to catch missing dependencies) or ABI-impacting things, but may not wait otherwise. 10 hours is an overkill, but it did pass this time.

File: mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp [hunk 2]
```diff
@@ -3269,22 +3269,24 @@ DiagnosedSilenceableFailure transform::FlattenElementwiseLinalgOp::applyToOne(
...
+    return mlir::emitSilenceableFailure(target->getLoc())
+           << "only elementwise flattening is supported";
+
   // If rank <= 1, do nothing
   if (target.getNumLoops() <= 1) {
     results.push_back(target);
     return DiagnosedSilenceableFailure::success();
   }
+
+  // Attempt to flatten all dims to one
```
Comment (ftynse) [inline mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:3282]: Nit: end comments with a period.

File: mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp [hunk 3]
```diff
@@ -3269,22 +3269,24 @@ DiagnosedSilenceableFailure transform::FlattenElementwiseLinalgOp::applyToOne(
...
   ReassociationIndices reassociation(target.getNumLoops());
   std::iota(reassociation.begin(), reassociation.end(), 0);
   auto maybeFlattened =
       collapseOpIterationDims(target, reassociation, rewriter);
   if (failed(maybeFlattened))
-    return emitDefaultSilenceableFailure(target);
+    return mlir::emitSilenceableFailure(target->getLoc())
+           << "Attempted to flatten, but failed";
```
Comment (ftynse) [inline mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:3289]: Nit: error messages should start with a lowercase letter.

**Review by ftynse — APPROVED** [review]

**Comment by ftynse** [comment]
> need to fix this one too https://github.com/llvm/llvm-project/pull/86146/files#diff-61bfc4a79549b25af3e46ff8002c9a9aea61fb537b565435ba0957e2b53b9adaR3286. just need to figure out a test case to trigger it

Thanks!

**Pull request merged**